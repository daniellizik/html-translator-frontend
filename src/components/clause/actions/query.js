import * as constants from '../constants'

export const remove = (clauseIndex, ruleIndex) => ({
  type: constants.QUERY_REMOVE,
  clauseIndex,
  ruleIndex,
  ruleType: 'QUERY'
})

export const add = (clauseIndex) => ({
  type: constants.QUERY_ADD,
  clauseIndex
})

// for async search
export const changeInit = () => ({ type: constants.QUERY_CHANGE_INIT })

export const changeError = (error) => ({
  type: constants.QUERY_CHANGE_ERROR,
  error
})

export const changeDone = (props) => ({
  ...props,
  type: constants.QUERY_CHANGE_DONE
})

export const changeTargetValue = (targetValue, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_TARGET_VALUE,
  targetValue,
  clauseIndex,
  ruleIndex
})

export const changeTargetValueAsync = (tv, ci, r) => (dispatch) => {
  
}

export const changeRule = (rule, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_RULE,
  rule,
  clauseIndex,
  ruleIndex
})

export const changeTarget = (target, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_TARGET,
  target,
  clauseIndex,
  ruleIndex
})

export const changeRuleValue = (ruleValue, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_RULE_VALUE,
  ruleValue,
  clauseIndex,
  ruleIndex
})

export const changeRuleValueFlags = (ruleValueFlags, clauseIndex, ruleIndex) => ({
  type: constants.QUERY_CHANGE_RULE_VALUE_FLAGS,
  ruleValueFlags,
  clauseIndex,
  ruleIndex
})
