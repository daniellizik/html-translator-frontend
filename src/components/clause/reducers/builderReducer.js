import * as constants from '../constants'
import { defaultClause } from '../settings/config'
import { bindConstantsToReducers } from '~/src/util'
import { reduceView, reduceClauses, mutationDenormalizer } from './util'

export default bindConstantsToReducers({
  [constants.BUILDER_REMOVE_ALL_CLAUSES]: (state, action) => ({
    ...state,
    activeClause: -1,
    clauses: []
  }),
  [constants.BUILDER_HIDE_ALL_MUTATIONS]: (state, action) => ({
    ...state,
    activeClause: 0,
    slave: {
      ...state.slave,
      mutated: state.slave.list.list,
      view: []
    }
  }),
  [constants.BUILDER_VIEW_ALL_MUTATIONS]: (state, action) => {
    const mutated = state
      .clauses
      .reduce((list, clause) => mutationDenormalizer(clause, list), state.slave.list.list)
    return {
      ...state,
      activeClause: -1,
      slave: {
        ...state.slave,
        mutated,
        view: mutated
      }
    }
  }
})