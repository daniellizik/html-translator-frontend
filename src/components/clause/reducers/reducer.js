import * as clauseConstants from '../constants'
import * as sourceSetterConstants from '~/src/containers/sourceSetter/constants'
import { defaultQuery, defaultClause, mutator } from '../config'
import queryReducer from './queryReducer'
import mutationReducer from './mutationReducer'
import clauseReducer from './clauseReducer'
import { reduceView } from './util'

// by default, adjusting a clause's inputs will set it to active
// those are set implicitly by other actions besides CLAUSE_ACTIVATE
// this root-level reducer is set before the default one for clause
// in rootReducer, since everything clause reducer does depends on 
// activeClause prop (basically)
export function activeClause(state, action) {
  let nextState = state
  if (action.clauseIndex || action.type === clauseConstants.CLAUSE_ACTIVATE)
    nextState = {
      ...state,
      activeClause: action.clauseIndex
    }
  return nextState
}

export default function(state, action) {
  let nextState = state
  if (action.type.indexOf('CLAUSE_') > -1)
    nextState = clauseReducer(state, action)
  else if (action.type.indexOf('QUERY_') > -1)
    nextState = queryReducer(state, action)
  else if (action.type.indexOf('MUTATION_') > -1)
    nextState = mutationReducer(state, action)
  return nextState
}
