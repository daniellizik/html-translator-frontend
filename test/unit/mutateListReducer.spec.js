import { mutateList } from '~/src/components/clause/reducers/util'
import { modifyingTextQueries as story } from '~/test/storyFixtures/query'

describe('mutate list', () => {
  const mutated = mutateList(story[2].slave.list.list, story[2].clauses[0].view)
  it('should mutate a list', () => {
    expect(mutated.length).toBe(story[2].slave.list.list.length)
  })
  it('should give node in view view: true prop', () => {
    expect(mutated.filter(m => m.view === true).length).toBe(6)
  })
})