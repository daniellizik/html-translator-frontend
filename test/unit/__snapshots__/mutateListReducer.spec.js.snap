exports[`mutate list should mutate list 1`] = `
Object {
  "list": Array [
    Object {
      "attrs": Array [],
      "depth": 0,
      "id": 0,
      "nodeName": "html",
      "parent": -1,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 1,
      "id": 1,
      "nodeName": "head",
      "parent": 0,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 2,
      "nodeName": "#text",
      "parent": 1,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 2,
      "nodeName": "#text",
      "parent": 1,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 3,
      "nodeName": "title",
      "parent": 2,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 4,
      "nodeName": "#text",
      "parent": 3,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 4,
      "nodeName": "#text",
      "parent": 3,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 3,
      "nodeName": "title",
      "parent": 2,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 5,
      "nodeName": "#text",
      "parent": 4,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 5,
      "nodeName": "#text",
      "parent": 4,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 1,
      "id": 1,
      "nodeName": "head",
      "parent": 0,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 1,
      "id": 6,
      "nodeName": "#text",
      "parent": 5,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 1,
      "id": 6,
      "nodeName": "#text",
      "parent": 5,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 1,
      "id": 7,
      "nodeName": "body",
      "parent": 6,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 8,
      "nodeName": "#text",
      "parent": 7,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 8,
      "nodeName": "#text",
      "parent": 7,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 9,
      "nodeName": "input",
      "parent": 8,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 9,
      "nodeName": "input",
      "parent": 8,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 10,
      "nodeName": "#text",
      "parent": 9,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 10,
      "nodeName": "#text",
      "parent": 9,
      "value": "",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "b",
          "value": "some food",
        },
        Object {
          "name": "cat",
          "value": "cat a taco power",
        },
      ],
      "depth": 2,
      "id": 11,
      "nodeName": "h2",
      "parent": 10,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 12,
      "nodeName": "#text",
      "parent": 11,
      "value": "cat-a",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 12,
      "nodeName": "#text",
      "parent": 11,
      "value": "cat-a",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "b",
          "value": "some food",
        },
        Object {
          "name": "cat",
          "value": "cat a taco power",
        },
      ],
      "close": true,
      "depth": 2,
      "id": 11,
      "nodeName": "h2",
      "parent": 10,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 13,
      "nodeName": "#text",
      "parent": 12,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 13,
      "nodeName": "#text",
      "parent": 12,
      "value": "",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "q",
          "value": "something",
        },
      ],
      "depth": 2,
      "id": 14,
      "nodeName": "div",
      "parent": 13,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 15,
      "nodeName": "#text",
      "parent": 14,
      "value": "cat-b",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 15,
      "nodeName": "#text",
      "parent": 14,
      "value": "cat-b",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "q",
          "value": "something",
        },
      ],
      "close": true,
      "depth": 2,
      "id": 14,
      "nodeName": "div",
      "parent": 13,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 16,
      "nodeName": "#text",
      "parent": 15,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 16,
      "nodeName": "#text",
      "parent": 15,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 17,
      "nodeName": "div",
      "parent": 16,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 18,
      "nodeName": "#text",
      "parent": 17,
      "value": "cat-c",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 18,
      "nodeName": "#text",
      "parent": 17,
      "value": "cat-c",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 17,
      "nodeName": "div",
      "parent": 16,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 19,
      "nodeName": "#text",
      "parent": 18,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 19,
      "nodeName": "#text",
      "parent": 18,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 20,
      "nodeName": "div",
      "parent": 19,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 21,
      "nodeName": "#text",
      "parent": 20,
      "value": "some random text",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 21,
      "nodeName": "#text",
      "parent": 20,
      "value": "some random text",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 22,
      "nodeName": "span",
      "parent": 21,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 23,
      "nodeName": "#text",
      "parent": 22,
      "value": "cat burrito",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 23,
      "nodeName": "#text",
      "parent": 22,
      "value": "cat burrito",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 22,
      "nodeName": "span",
      "parent": 21,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 24,
      "nodeName": "#text",
      "parent": 23,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 24,
      "nodeName": "#text",
      "parent": 23,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 25,
      "nodeName": "span",
      "parent": 24,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 26,
      "nodeName": "#text",
      "parent": 25,
      "value": "cat taco",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 26,
      "nodeName": "#text",
      "parent": 25,
      "value": "cat taco",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 25,
      "nodeName": "span",
      "parent": 24,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 27,
      "nodeName": "#text",
      "parent": 26,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 27,
      "nodeName": "#text",
      "parent": 26,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 28,
      "nodeName": "span",
      "parent": 27,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 29,
      "nodeName": "#text",
      "parent": 28,
      "value": "cat empanada",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 29,
      "nodeName": "#text",
      "parent": 28,
      "value": "cat empanada",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 28,
      "nodeName": "span",
      "parent": 27,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 30,
      "nodeName": "#text",
      "parent": 29,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 30,
      "nodeName": "#text",
      "parent": 29,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 31,
      "nodeName": "span",
      "parent": 30,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 32,
      "nodeName": "#text",
      "parent": 31,
      "value": "cat taquito",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 32,
      "nodeName": "#text",
      "parent": 31,
      "value": "cat taquito",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 31,
      "nodeName": "span",
      "parent": 30,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 33,
      "nodeName": "#text",
      "parent": 32,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 33,
      "nodeName": "#text",
      "parent": 32,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 34,
      "nodeName": "span",
      "parent": 33,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 35,
      "nodeName": "#text",
      "parent": 34,
      "value": "cat fajita",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 35,
      "nodeName": "#text",
      "parent": 34,
      "value": "cat fajita",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 34,
      "nodeName": "span",
      "parent": 33,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 36,
      "nodeName": "#text",
      "parent": 35,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 36,
      "nodeName": "#text",
      "parent": 35,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 37,
      "nodeName": "div",
      "parent": 36,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 38,
      "nodeName": "#text",
      "parent": 37,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 38,
      "nodeName": "#text",
      "parent": 37,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 39,
      "nodeName": "div",
      "parent": 38,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 40,
      "nodeName": "#text",
      "parent": 39,
      "value": "cat burrito bacon",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 5,
      "id": 40,
      "nodeName": "#text",
      "parent": 39,
      "value": "cat burrito bacon",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 39,
      "nodeName": "div",
      "parent": 38,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 41,
      "nodeName": "#text",
      "parent": 40,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 41,
      "nodeName": "#text",
      "parent": 40,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 42,
      "nodeName": "div",
      "parent": 41,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 43,
      "nodeName": "#text",
      "parent": 42,
      "value": "cat burrito avocado",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 5,
      "id": 43,
      "nodeName": "#text",
      "parent": 42,
      "value": "cat burrito avocado",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 42,
      "nodeName": "div",
      "parent": 41,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 44,
      "nodeName": "#text",
      "parent": 43,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 44,
      "nodeName": "#text",
      "parent": 43,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 45,
      "nodeName": "div",
      "parent": 44,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 46,
      "nodeName": "#text",
      "parent": 45,
      "value": "cat burrito ice         cream",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 5,
      "id": 46,
      "nodeName": "#text",
      "parent": 45,
      "value": "cat burrito ice         cream",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 45,
      "nodeName": "div",
      "parent": 44,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 47,
      "nodeName": "#text",
      "parent": 46,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 47,
      "nodeName": "#text",
      "parent": 46,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 48,
      "nodeName": "div",
      "parent": 47,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 49,
      "nodeName": "#text",
      "parent": 48,
      "value": "cat burrito ice         spleen",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 5,
      "id": 49,
      "nodeName": "#text",
      "parent": 48,
      "value": "cat burrito ice         spleen",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 48,
      "nodeName": "div",
      "parent": 47,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 50,
      "nodeName": "#text",
      "parent": 49,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 50,
      "nodeName": "#text",
      "parent": 49,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 51,
      "nodeName": "div",
      "parent": 50,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 52,
      "nodeName": "#text",
      "parent": 51,
      "value": "cat burrito spinache",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 5,
      "id": 52,
      "nodeName": "#text",
      "parent": 51,
      "value": "cat burrito spinache",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 51,
      "nodeName": "div",
      "parent": 50,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 53,
      "nodeName": "#text",
      "parent": 52,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 53,
      "nodeName": "#text",
      "parent": 52,
      "value": "",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "data-cat",
          "value": "burrito",
        },
      ],
      "depth": 4,
      "id": 54,
      "nodeName": "div",
      "parent": 53,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 55,
      "nodeName": "#text",
      "parent": 54,
      "value": "cat burrito cheese",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 5,
      "id": 55,
      "nodeName": "#text",
      "parent": 54,
      "value": "cat burrito cheese",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "data-cat",
          "value": "burrito",
        },
      ],
      "close": true,
      "depth": 4,
      "id": 54,
      "nodeName": "div",
      "parent": 53,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 56,
      "nodeName": "#text",
      "parent": 55,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 4,
      "id": 56,
      "nodeName": "#text",
      "parent": 55,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 37,
      "nodeName": "div",
      "parent": 36,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 57,
      "nodeName": "#text",
      "parent": 56,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 3,
      "id": 57,
      "nodeName": "#text",
      "parent": 56,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 20,
      "nodeName": "div",
      "parent": 19,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 58,
      "nodeName": "#text",
      "parent": 57,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 2,
      "id": 58,
      "nodeName": "#text",
      "parent": 57,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 1,
      "id": 7,
      "nodeName": "body",
      "parent": 6,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "close": true,
      "depth": 0,
      "id": 0,
      "nodeName": "html",
      "parent": -1,
      "value": undefined,
    },
  ],
  "open": Array [
    Object {
      "attrs": Array [],
      "depth": 0,
      "id": 0,
      "nodeName": "html",
      "parent": -1,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 1,
      "id": 1,
      "nodeName": "head",
      "parent": 0,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 2,
      "nodeName": "#text",
      "parent": 1,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 3,
      "nodeName": "title",
      "parent": 2,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 4,
      "nodeName": "#text",
      "parent": 3,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 5,
      "nodeName": "#text",
      "parent": 4,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 1,
      "id": 6,
      "nodeName": "#text",
      "parent": 5,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 1,
      "id": 7,
      "nodeName": "body",
      "parent": 6,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 8,
      "nodeName": "#text",
      "parent": 7,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 9,
      "nodeName": "input",
      "parent": 8,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 10,
      "nodeName": "#text",
      "parent": 9,
      "value": "",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "b",
          "value": "some food",
        },
        Object {
          "name": "cat",
          "value": "cat a taco power",
        },
      ],
      "depth": 2,
      "id": 11,
      "nodeName": "h2",
      "parent": 10,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 12,
      "nodeName": "#text",
      "parent": 11,
      "value": "cat-a",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 13,
      "nodeName": "#text",
      "parent": 12,
      "value": "",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "q",
          "value": "something",
        },
      ],
      "depth": 2,
      "id": 14,
      "nodeName": "div",
      "parent": 13,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 15,
      "nodeName": "#text",
      "parent": 14,
      "value": "cat-b",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 16,
      "nodeName": "#text",
      "parent": 15,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 17,
      "nodeName": "div",
      "parent": 16,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 18,
      "nodeName": "#text",
      "parent": 17,
      "value": "cat-c",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 19,
      "nodeName": "#text",
      "parent": 18,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 20,
      "nodeName": "div",
      "parent": 19,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 21,
      "nodeName": "#text",
      "parent": 20,
      "value": "some random text",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 22,
      "nodeName": "span",
      "parent": 21,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 23,
      "nodeName": "#text",
      "parent": 22,
      "value": "cat burrito",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 24,
      "nodeName": "#text",
      "parent": 23,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 25,
      "nodeName": "span",
      "parent": 24,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 26,
      "nodeName": "#text",
      "parent": 25,
      "value": "cat taco",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 27,
      "nodeName": "#text",
      "parent": 26,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 28,
      "nodeName": "span",
      "parent": 27,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 29,
      "nodeName": "#text",
      "parent": 28,
      "value": "cat empanada",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 30,
      "nodeName": "#text",
      "parent": 29,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 31,
      "nodeName": "span",
      "parent": 30,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 32,
      "nodeName": "#text",
      "parent": 31,
      "value": "cat taquito",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 33,
      "nodeName": "#text",
      "parent": 32,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 34,
      "nodeName": "span",
      "parent": 33,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 35,
      "nodeName": "#text",
      "parent": 34,
      "value": "cat fajita",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 36,
      "nodeName": "#text",
      "parent": 35,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 37,
      "nodeName": "div",
      "parent": 36,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 38,
      "nodeName": "#text",
      "parent": 37,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 39,
      "nodeName": "div",
      "parent": 38,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 40,
      "nodeName": "#text",
      "parent": 39,
      "value": "cat burrito bacon",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 41,
      "nodeName": "#text",
      "parent": 40,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 42,
      "nodeName": "div",
      "parent": 41,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 43,
      "nodeName": "#text",
      "parent": 42,
      "value": "cat burrito avocado",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 44,
      "nodeName": "#text",
      "parent": 43,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 45,
      "nodeName": "div",
      "parent": 44,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 46,
      "nodeName": "#text",
      "parent": 45,
      "value": "cat burrito ice         cream",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 47,
      "nodeName": "#text",
      "parent": 46,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 48,
      "nodeName": "div",
      "parent": 47,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 49,
      "nodeName": "#text",
      "parent": 48,
      "value": "cat burrito ice         spleen",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 50,
      "nodeName": "#text",
      "parent": 49,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 51,
      "nodeName": "div",
      "parent": 50,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 52,
      "nodeName": "#text",
      "parent": 51,
      "value": "cat burrito spinache",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 53,
      "nodeName": "#text",
      "parent": 52,
      "value": "",
    },
    Object {
      "attrs": Array [
        Object {
          "name": "data-cat",
          "value": "burrito",
        },
      ],
      "depth": 4,
      "id": 54,
      "nodeName": "div",
      "parent": 53,
      "value": undefined,
    },
    Object {
      "attrs": Array [],
      "depth": 5,
      "id": 55,
      "nodeName": "#text",
      "parent": 54,
      "value": "cat burrito cheese",
    },
    Object {
      "attrs": Array [],
      "depth": 4,
      "id": 56,
      "nodeName": "#text",
      "parent": 55,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 3,
      "id": 57,
      "nodeName": "#text",
      "parent": 56,
      "value": "",
    },
    Object {
      "attrs": Array [],
      "depth": 2,
      "id": 58,
      "nodeName": "#text",
      "parent": 57,
      "value": "",
    },
  ],
}
`;
